### ISSUE TITLE
Complete warnings for AOT unsafe interop

### ISSUE THREAD

**Issue by jkotas** [dotnet/runtime#74697]
Warnings for AOT unsafe interop are incomplete best effort currently: https://github.com/dotnet/runtime/pull/74630#discussion_r955966809

- [ ] Generate warnings for unsafe interop in marshalled structs
- [ ] Generate warnings for AOT unsafe marshallers like AsAny and LayoutClass 
- [ ] Use actual marshaller kind to produce the warning instead of the approximation at https://github.com/dotnet/runtime/blob/2201016c1e13bdb9abf49e2e38cadf4ee0568df2/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Dataflow/ReflectionMethodBodyScanner.cs#L566-L643
- [ ] Steer people towards P/Invoke source generator by the warning text

**Comment by MichalStrehovsky** [comment]
> Steer people towards P/Invoke source generator by the warning text

If we tell people to use the source generator for AsAny/LayoutClass/System.Delegate marshalling, the source generator will then tell them it's unsupported. It will shift the blame, but the end result for the user is the same if we don't send them on a hike first.

**Comment by jkotas** [comment]
Yes, it is the case with retrofitting existing AOT unsafe code.

Steering people towards source generated interop is more interesting for new code. Source generated interop has natural AOT safety guiderails and prevents the AOT unsafe code to be written in the first place.

**Comment by hez2010** [comment]
> towards P/Invoke source generator

Is source generated P/Invokes support direct P/Invoke and static linking?

**Comment by jkotas** [comment]
> Do source generated P/Invokes support direct P/Invoke and static linking?

Yes. This aspect is not different between source generated and built-in marshalling.

All P/Invokes (except a few outliers) in this repo have been switched to be source generated, and native aot publishing uses direct P/Invokes and static linking extensively.

**Comment by MichalStrehovsky** [comment]
> Use actual marshaller kind to produce the warning instead of the approximation at

This logic is shared with IL Linker and IL Linker doesn't have access to this unfortunately.
